Add resize layer buttons

Mxchaeltrxn opened the pull request:
Addresses #15.

I'm also unsure about styling while we're using text instead of an icon.
Otherwise, I think it's fine.

haydn commented on src/Editor.js:
```diff
@@ -455,6 +455,29 @@ const Editor = () => {
+          }
+
+          return layer;
+        })
```
This function is currently mutating the state, which we want to avoid ([React docs for state](https://reactjs.org/docs/react-component.html#state), [React docs about performance](https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data) and [an article giving a good explanation](https://www.freecodecamp.org/news/handling-state-in-react-four-immutable-approaches-to-consider-d1f5c00249d5/)).

This is happening because both `updatedLayer` and `layer` are references to the same object (`updatedLayer === layer`) — any changes made on `updatedLayer` are changes on `layer`.

To avoid mutating the state, we need to create a new object and copy the values across. To do that we can use the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax):

```js
const updatedLayer = { ...layer };
```

haydn commented on src/Editor.js:
```diff
@@ -1234,6 +1257,97 @@ const Editor = () => {
+                    ) {
+                      widestSelectedLayer = layer;
+                    }
+                  });
```
We can refactor this using a [filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) and [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce) to be a little more methodical and functional:

```js
const allSelectedLayers = doc.layers.filter(layer => selection.has(layer.id));
const widestSelectedLayer = allSelectedLayers.reduce(
  (widestFoundSoFar, layer) => (
    widestFoundSoFar.x2 - widestFoundSoFar.x1 < layer.x2 - layer.x1
      ? layer
      : widestFoundSoFar
  ), // The return value of this function will be the value of the widestFoundSoFar argument when this function is called again in the next iteration
  { x1: 0, x2: 0 } // This is the initial value of widestFoundSoFar in the first iteration
);
```

haydn replied:
I also think it'd be good to move this into the `resizeLayerByExtreme` function so that we can use it something like this:

```js
resizeLayerByExtreme(layers, "widest", layer => selection.has(layer.id));
```

That way the `resizeLayerByExtreme` function can be responsible for both identifying the widest/narrowest/tallest/shortest layer as well as resizing the other layers.

Alternatively, this function could be called something like `setLayerDimensions` that we use like this:

```js
const layersSetToSameX = setLayerDimensions(layers, { x1: 100, x2: 200 });
const layersSetToSameY = setLayerDimensions(layers, { y1: 50, y2: 100 });
const layersSetToSameY2 = setLayerDimensions(layers, { y2: 400 });
```

The problem I'm trying to solve here is that `resizeLayerByExtreme` is a bit of a misleading name, because the `extreme` argument that is passed to it is really just deciding if it considers the `x` or `y` axis, but the function doesn't really do anything that considers extremes.

Mxchaeltrxn replied:
This makes a lot of sense to me. I don't really get why the third argument is passed in as an argument or how it'll be used in the below snippet.
```
resizeLayerByExtreme(layers, "widest", layer => selection.has(layer.id));
```

haydn replied:
It makes it easier to use in this pattern (you'll notice a lot of the other functions in the `layers.js` file have similar signatures):

```js
setState(current => ({
  ...current,
  doc: {
    ...current.doc,
    layers: resizeLayerByExtreme(current.doc.layers, "widest", layer => selection.has(layer.id))
  }
}));
```

You can pass it an array of layers and a [predicate](https://en.wikipedia.org/wiki/Predicate_(mathematical_logic)) so that it can return an array of all the same layers (in the same order) with just the desired ones updated.

Mxchaeltrxn replied:
Thanks for the explanations

Mxchaeltrxn commented:
I've addressed all your comments. I've also put the switch statement inside the reduce function on line 473, which caused me to name the accumulator 'mostExtremeDimensionsFoundSoFar'. I'm unsure how clear that name is (which is a bad sign) but I think it's much cleaner than putting the switch statement outside and the reduce functions inside.

Mxchaeltrxn commented:
Noted everything you said. Thanks a lot for the detail. I addressed what you said and I removed my old updateLayersDimensions function because it wasn't needed anymore. I also added tests for the new functions made. 

I'm assuming for the predicate, we use the predicate that we use when calling the function. I can add tests for the other functions as well because I notice they're missing tests for the predicate as well.

haydn commented on src/Editor.js:
```diff
@@ -1308,6 +1313,43 @@ const Editor = () => {
+                  {label}
+                </Button>
+              ))}
+            </div>
```
Nice! 😍

haydn commented:
@Mxchaeltrxn Nice work! Very tidy. 😄 I've created a new issue to track those missing predicate tests (#75).
